### PR TITLE
Refactor: Create autofill input for pickers

### PR DESCRIPTION
**Pull request by joschect**
Greatly simplify the picker code by spinning out the input autofill into it's own component.

### CODE REVIEW

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  public componentDidUpdate() {
+    let { value } = this.state;
+    let { suggestedDisplayValue } = this.props;
+    let differenceIndex = 0;
+    if (value && suggestedDisplayValue && this._doesTextStartWith(suggestedDisplayValue, value)) {
+      while (differenceIndex < value.length && value[differenceIndex].toLocaleLowerCase() === suggestedDisplayValue[differenceIndex].toLocaleLowerCase()) {
+        differenceIndex++;
+      }
+      if (differenceIndex > 0) {
+        this._inputElement.setSelectionRange(differenceIndex, suggestedDisplayValue.length, 'backward');
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: const dis

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    let {
+      ariaDescription,
+      ariaActiveDescendant,
+      ariaOwns,
+      ariaExpanded,
+      ariaHasPopup,
+      suggestedDisplayValue
+    } = this.props;
+
+    let displayValue = this.state.value;
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: value

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    if (this._doesTextStartWith(suggestedDisplayValue, value)) {
+      displayValue = suggestedDisplayValue;
+    }
+    const nativeProps = getNativeProps(this.props, inputProperties);
+    return React.createElement(tag,
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: return `<input ...`

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+import { getNativeProps, inputProperties} from '../../../utilities/properties';
+import { autobind } from '../../../utilities/autobind';
+import { KeyCodes } from '../../../utilities/KeyCodes';
+
+export interface IBaseAutoFillState {
+  value?: string;
+}
+
+const tag = 'input';
+export class BaseAutoFill extends BaseComponent<IBaseAutoFillProps, IBaseAutoFillState> {
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: missing IBaseAutoFill ? That has your publics defined maybe?

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  @autobind
+  private _onKeyDown(ev: React.KeyboardEvent<HTMLElement>) {
+    switch (ev.which) {
+      case KeyCodes.backspace:
+        this._onBackspace(ev);
+        break;
+    }
+  }
+
+  private _onBackspace(ev: React.KeyboardEvent<HTMLElement>) {
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: maybe _handleBackspace

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    let { value } = this.state;
+    if (ev.target === this._inputElement) {
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: is this necessary?

File: src/components/pickers/AutoFill/BaseAutoFill.tsx [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      if (value && value.length > 0) {
+        this._updateValue(value.substring(0, value.length - 1));
```
Comment (dzearing) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: substr or substring? Look it up :)

File: src/components/pickers/BasePicker.tsx [hunk 8]
```diff
@@ -256,16 +228,16 @@ export class BasePicker<T, P extends IBasePickerProps<T>> extends BaseComponent<
   }
 
   @autobind
-  protected onInputFocus(ev: React.FocusEvent<HTMLElement>) {
+  protected onInputFocus(ev: React.FocusEvent<HTMLInputElement | BaseAutoFill>) {
```
Comment (dzearing) [inline src/components/pickers/BasePicker.tsx]: maybe add a test?

File: src/components/pickers/AutoFill/BaseAutoFill.tsx (hunk 7 above)
Comment (joschect) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: So they are different, but which one to use isn't really clear. Substr is substr(start, length) and substring is substring(startindex, endindex). Substr supports a negative start and will go from the end of the string. So with substr you can do ('hello there').substr(-4, 2) will produce "he".

File: src/components/pickers/AutoFill/BaseAutoFill.tsx (hunk 6 above)
Comment (joschect) [inline src/components/pickers/AutoFill/BaseAutoFill.tsx]: Nope

**Pull request merged**